worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/ArgumentsManager.java

```diff
@@ -177,5 +198,51 @@ public void checkHealth() {
             }
         }
     }
+    
+    private boolean isPoisonDocument(final Document document) {
+        // A poison document is a document that a downstream worker has redirected back
+        // to the workflow worker. A document is considered poison if:
+        //
+        // 1. It has a a non-empty CAF_WORKFLOW_SETTINGS field.
+        //
+        // 2. It does NOT have a 'tenantId' inside it's custom data.
+        //
+        // Point (2) is important, it's not enough to just check for a non-empty 
+        // CAF_WORKFLOW_SETTINGS field, since that makes it possible for a rogue agent to
+        // stage documents that already contain CAF_WORKFLOW_SETTINGS field, which could
+        // possibly be used to write to, or delete from, another tenant's index.
+        //
+        // The tenantId that is passed in custom data cannot be controlled by a rogue agent 
+        // and is therefore trustworthy, so if that is checked as well, it gives some 
+        // confidence that this is really a poison document, and not a document staged by
+        // a rogue agent, and as such we can safely use the CAF_WORKFLOW_SETTINGS present
+        // in the document and trust that they are valid.
+        return document.getField("CAF_WORKFLOW_SETTINGS").hasValues() && 
+            (document.getTask().getCustomData("tenantId") == null);
```

Comment:
As discussed, let's change this to look for "workflowName" rather than "tenantId", given that tenant is not a core concept to the worker.